### PR TITLE
Automatically pan initial map based on current location

### PR DESCRIPTION
**Pull request by MaciejSWM**
### Details

Thread: https://expensify.slack.com/archives/C01GTK53T8Q/p1698404342523299

### Fixed Issues

$ https://github.com/Expensify/App/issues/22704
PROPOSAL: 

## Web

There are three possibilities when testing the Web, depending on user's location permissions:

**When page `Wants to know your location`:**
image
image
1. Logout. Reset page's permissions so that browser can ask you for Location permission. Login and go to Request Money -> Distance. Browser should ask if you want to grant location access. Map should stay in pending state until you choose something. Grant the permission. Map should navigate straight to your current location within 1-2 seconds.
2. Do step 1. again. This time don't grant the permission. Map should take you to the default San Francisco location.

**When location access allowed:**
image
1. Logout. Then login and go to Request Money -> Distance. You should see the map in pending state. Once map is fully loaded, it should take you straight to your location. There should be no animation, just land straight in the current location.
2. In `DistanceRequestFooter.js` add the following:
```
    useEffect(() => {
        Onyx.set(ONYXKEYS.USER_LOCATION, { longitude: 12, latitude: 12 })
    }, [])
```
Then open Request Money -> Distance. You should see the map initially load the [12, 12] coords, and after a second or so it should take you to your current location, once your location is obtained.
3. Do the same as in step 2., but before your current location is obtained, drag the map with the mouse. This should keep you at [12, 12] coords and should stop the map from auto-panning to your current location.

**When location access is disabled:**
image
- Logout. Then login and go to Request Money -> Distance. You should land in the default location, which is San Francisco.
- Refresh the page. You should land straight in the San Francisco location.
- Enable the location and refresh the page to cache the current location in ONYX. Then disable the location and refresh the page. You should land in this cached location instead of the default San Francisco location.

# Desktop

You have to control the location permissions from the system settings:
image

Test all the scenarios (enabled/ask/disabled) similarly to web.

# Mobile
Control the location permissions through your phone settings:
image

Test all the scenarios (enabled/ask/disabled) similarly to web.

### Offline tests

Map won't load in offline mode. Browser won't ask about granting location permissions.

### QA Steps

- Same as `Tests`

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
        - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

(Ignore console warnings - I added them for easier debugging)

https://github.com/Expensify/App/assets/85628901/819bda7d-c084-47dc-b45f-353dc053c0d9

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/85628901/60ad5911-7b7e-4b89-bfd8-b8862540d12f

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/85628901/5c0ca521-d2a2-4ca8-b46d-7829ca7696e2

https://github.com/Expensify/App/assets/85628901/2b8eefde-f05d-4df7-b3f1-1825872045f0

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/85628901/34d4b2ca-3be9-4515-916f-b0caee88de24

https://github.com/Expensify/App/assets/85628901/c99eaa77-3283-4fc1-abd5-a192b584bbc7

https://github.com/Expensify/App/assets/85628901/954204a8-951e-47ed-8100-d23ee9bbe499

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/85628901/ee29e9fe-0954-4d3c-a77a-c099eb11fcb0

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/85628901/3cff82e0-27a5-493a-9f83-71df394f3538

https://github.com/Expensify/App/assets/85628901/77c82f1e-2567-4db9-b089-da5ca7120ce8

</details>

### CODE REVIEW

File: src/hooks/useLocalize.js [hunk 1]
```diff
@@ -1,6 +1,24 @@
...
+ * @property {(digit: string) => string} toLocaleDigit Converts the provided digit to the locale digit.
+ * @property {(localeDigit: string) => string} fromLocaleDigit Reverses the operation of `toLocaleDigit`, taking a locale-specific digit and returning the equivalent in the standard number system.
+ * @property {string} preferredLocale The preferred locale value.
+ */
+
+/**
+ * Hook to access the localization context which provides multiple utility functions and locale.
+ *
+ * @returns {LocalizationContext} The localization context
+ */
```
Comment (MaciejSWM) [inline src/hooks/useLocalize.js:21]: Confirmed by @blazejkustra that this is the way we can handle it for now until `useLocalize` is migrated to TS.

Without this type definition, usage of `useLocalize` in `MapView.tsx` would throw TS errors

File: src/components/DistanceMapView/distanceMapViewPropTypes.js [hunk 2]
```diff
@@ -2,7 +2,7 @@ import PropTypes from 'prop-types';
 
 const propTypes = {
     // Public access token to be used to fetch map data from Mapbox.
-    accessToken: PropTypes.string.isRequired,
+    accessToken: PropTypes.string,
```
Comment (mananjadhav) [inline src/components/DistanceMapView/distanceMapViewPropTypes.js:5]: Why do we need to remove `isRequired`? and if we're removing, then shouldn't we also set some default value?

**Comment by mananjadhav** [comment]
@MaciejSWM Started reviewing, can you please merge the latest main and resolve conflicts?

**Comment by MaciejSWM** [comment]
@mananjadhav conflicts resolved

File: src/components/MapView/MapView.web.tsx [hunk 3]
```diff
@@ -77,40 +148,55 @@ const MapView = forwardRef<MapViewHandle, MapViewProps>(
...
-            </View>
+            <>
+                {!isOffline && Boolean(accessToken) && Boolean(currentPosition) ? (
+                    <View
+                        style={style}
+                        // eslint-disable-next-line react/jsx-props-no-spreading
+                        {...responder.panHandlers}
+                    >
+                        <Map
+                            onDrag={() => setUserInteractedWithMap(true)}
```
Comment (WoLewicki) [inline src/components/MapView/MapView.web.tsx:159]: Are there any other options for user to interact with the map?

Comment (MaciejSWM) [reply src/components/MapView/MapView.web.tsx:159]: I only care about user dragging the map - if he drags then I want to stop panning the map. If he clicks/presses/zooms or anything else, we are fine with panning the map. This made me think - maybe a better name is `setUserInteractedWithMap` -> `setUserHasDraggedMap`

Comment (WoLewicki) [reply src/components/MapView/MapView.web.tsx:159]: Yeah, sounds better this way

**Comment by dylanexpensify** [comment]
bumping this! What's left to get this to merged heaven?

**Comment by mananjadhav** [comment]
Checked the code. @MaciejSWM Can you please resolve the conflicts? So that I can finish this by tomorrow.

**Comment by MaciejSWM** [comment]
Conflicts resolved @mananjadhav

**Comment by dylanexpensify** [comment]
@mananjadhav can we get this reviewed again today?? TY!

**Comment by mananjadhav** [comment]
Thanks for the bump, I'll have this finished today.

File: src/components/DistanceMapView/distanceMapViewPropTypes.js (hunk 2 above)
Comment (mananjadhav) [reply src/components/DistanceMapView/distanceMapViewPropTypes.js:5]: @MaciejSWM This still seems not responded.

**Comment by mananjadhav** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/3069065/ddd917ec-5046-4144-9af6-fde1257cca2a

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/3069065/d2e8cc9e-ac32-4866-8a49-91002f6cf345

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/3069065/80826c10-f07c-4cb8-b533-3c7afc4f5faf

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/3069065/3975d7a3-1290-4345-b1c4-55aafee47410

https://github.com/Expensify/App/assets/3069065/d5da88fd-32fd-431e-8dd7-5232dab6913d

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/3069065/819e11f2-126a-413f-b442-f1b82a5ba768
</details>

**Comment by mananjadhav** [comment]
@MaciejSWM I had one issue testing with mobile web safari. The location shows momentarily and then goes black. Even when switching from default location to the user's location. Did you experience this? I have had this issue with my emulator earlier too, and I don't have a physical android device to test. But thought of clarifying with you before we merge.

**Review by mananjadhav — APPROVED** [review]

**Review by thienlnam — DISMISSED** [review]

**Comment by thienlnam** [comment]
Holding until we get those last couple of comments addressed

File: src/components/DistanceMapView/distanceMapViewPropTypes.js (hunk 2 above)
Comment (MaciejSWM) [reply src/components/DistanceMapView/distanceMapViewPropTypes.js]: Hmm I can't recall why. Seems like a mistake to me right now so I'm bringing it back. Thanks!

**Comment by MaciejSWM** [comment]
Hey @mananjadhav. I restored `isRequired` - looks like a mistake on my end. When it comes to testing I tested on Android emulator and on physical Android device with Chrome and never experienced black map. I'm checking one more time right now.

**Comment by MaciejSWM** [comment]
I'm having problems now to open it on any Android device - Emulator freezes once again because it runs out of memory. Physical device doesn't want to login to expensify because it's getting 429 Too many requests (tested ports 8082 and 8080)

**Comment by MaciejSWM** [comment]
@mananjadhav I confirm that it's still all good on my side:

https://github.com/Expensify/App/assets/85628901/4d78ced5-81ca-42d7-907f-0d621f0c03c9

**Comment by mananjadhav** [comment]
Thanks for confirming @MaciejSWM. It must be my emulator. 

@thienlnam All good here. I've also check off the checklist for the pending `propTypes` change.

**Review by mananjadhav — APPROVED** [review]

**Review by thienlnam — APPROVED** [review]

File: src/components/MapView/MapView.web.tsx [hunk 4]
```diff
@@ -34,7 +105,7 @@ const MapView = forwardRef<MapViewHandle, MapViewProps>(
             if (waypoints.length === 1) {
                 mapRef.flyTo({
                     center: waypoints[0].coordinate,
-                    zoom: 15,
+                    zoom: CONST.MAPBOX.DEFAULT_ZOOM,
```
Comment (shubham1206agra) [inline src/components/MapView/MapView.web.tsx:108]: Comment for reference

Comment (eVoloshchak) [reply src/components/MapView/MapView.web.tsx:108]: Just a heads up, this has caused an issue in https://github.com/Expensify/App/issues/32261

The zoom level was changed from 15 to 10 for web, but it's still 15 for other platforms. 
This caused an issue when the default zoom and zoom when you select a location is the same, so if the default location and the location you've selected are the same - the map isn't zoomed in
This was changed back to 15 in https://github.com/Expensify/App/pull/33234

**Pull request merged**